//
// Copyright 2018 - present the original author or authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
syntax = "proto3";

option go_package = "github.com/opencord/voltha/protos/go/omci";

package omci;

import "meta.proto";


message MibAttributeData {
    string name        = 1 [(voltha.access) = READ_ONLY];
    string value       = 2;
}

message MibInstanceData {
    uint32 instance_id      = 1 [(voltha.access) = READ_ONLY];
    string created          = 2;
    string modified         = 3;

    repeated MibAttributeData attributes = 4
    [(voltha.child_node) = {key: "name"}];
}

message MibClassData {
    uint32 class_id         = 1 [(voltha.access) = READ_ONLY];

    repeated MibInstanceData instances= 2
    [(voltha.child_node) = {key: "instance_id"}];
}

message ManagedEntity {
    uint32 class_id  = 1 [(voltha.access) = READ_ONLY];
    string name      = 2 [(voltha.access) = READ_ONLY];
}

message MessageType {
    uint32 message_type = 1 [(voltha.access) = READ_ONLY];
}

message MibDeviceData {
    string device_id        = 1 [(voltha.access) = READ_ONLY];
    string created          = 2;
    string last_sync_time   = 3;
    uint32 mib_data_sync    = 4;
    uint32 version          = 5;

    repeated MibClassData classes = 6
    [(voltha.child_node) = {key: "class_id"}];

    repeated ManagedEntity managed_entities = 7;
    repeated MessageType message_types = 8;
}

message OpenOmciEventType {
    enum OpenOmciEventType {
        state_change = 0; // A state machine has transitioned to a new state
    }
}

message OpenOmciEvent {
    OpenOmciEventType.OpenOmciEventType type = 1;

    string data = 2; // associated data, in json format
}
